### PR TITLE
initialize go mod for ecc; build vscc-plugin

### PR DESCRIPTION
**Pull request by MHBauer**
ecc has dependencies on both local modules of the top level fpc
module, as well as the ercc module.

Turned on the make target vscc-plugin.

**What this PR does / why we need it**:
related to #247 , allows the ecc-vscc plugin to build, as reported in chat. (was there an issue, or only reports of the problem?)

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

Potential TODO's here or in new issues.
 - [ ] update ercc go.mod
 - [ ] CI ercc and ecc builds to do -mod=readonly during CI

### CODE REVIEW

**Comment by MHBauer** [comment]
The dependency on the overall fpc module one directory above raises the level of some of the dependencies, so there will need to be changes in that module to sync up with fabric.

File: Makefile [hunk 1]
```diff
@@ -41,6 +41,8 @@ gotools:
 
 godeps: gotools
 	$(GO) mod download
+	cd ercc $(GO) mod download
```
Comment (g2flyer) [inline Makefile]: missing semicolon

Comment (MHBauer) [reply Makefile]: it's gone.

**Comment by g2flyer** [comment]
btw: i guess it's a go feature but it is surprising that now with modules a `go clean` triggered by  `make clean` can trigger downloading of go modules ...?!

**Comment by g2flyer** [comment]
Ok, just tried the PR on top of #307, did a `make clean all` and checked:
- does peer start with ecc-vscc enabled: answer no, failed launching peer with
```
panic: Error opening plugin at path ../../ecc/ecc-vscc.so: plugin.Open("../../ecc/ecc-vscc"): plugin was built with a different version of package github.com/hyperledger/fabric/common/errors
```
- does peer start with ecc-vscc diabled: answer no, ercc installation fails with
```docker build failed: Error returned from build: 1 "go: downloading github.com/hyperledger/fabric v2.0.1+incompatible
go: downloading github.com/hyperledger/fabric-chaincode-go v0.0.0-20200128192331-2d899240a7ed
go: downloading github.com/hyperledger/fabric-protos-go v0.0.0-20200124220212-e9cfc186ba7b
go: extracting github.com/hyperledger/fabric-protos-go v0.0.0-20200124220212-e9cfc186ba7b
go: downloading github.com/golang/protobuf v1.3.2
go: downloading google.golang.org/grpc v1.24.0
go: extracting github.com/hyperledger/fabric-chaincode-go v0.0.0-20200128192331-2d899240a7ed
go: extracting google.golang.org/grpc v1.24.0
go: downloading google.golang.org/genproto v0.0.0-20190819201941-24fa4b261c55
go: downloading golang.org/x/net v0.0.0-20200202094626-16171245cfb2
go: downloading golang.org/x/sys v0.0.0-20191120155948-bd437916bb0e
go: extracting github.com/golang/protobuf v1.3.2
go: extracting golang.org/x/net v0.0.0-20200202094626-16171245cfb2
go: downloading golang.org/x/text v0.3.2
go: extracting golang.org/x/sys v0.0.0-20191120155948-bd437916bb0e
go: extracting github.com/hyperledger/fabric v2.0.1+incompatible
go: downloading go.uber.org/zap v1.10.0
go: downloading github.com/sykesm/zap-logfmt v0.0.2
go: downloading github.com/pkg/errors v0.8.1
go: extracting google.golang.org/genproto v0.0.0-20190819201941-24fa4b261c55
go: extracting github.com/pkg/errors v0.8.1
go: extracting github.com/sykesm/zap-logfmt v0.0.2
go: extracting go.uber.org/zap v1.10.0
go: downloading go.uber.org/atomic v1.6.0
go: downloading go.uber.org/multierr v1.5.0
go: extracting golang.org/x/text v0.3.2
go: extracting go.uber.org/multierr v1.5.0
go: extracting go.uber.org/atomic v1.6.0
go: updates to go.mod needed, disabled by -mod=readonly
```

The `different version of package` error happened with both building fabric as `make native` as marcus said or based on `GO_TAGS=pluginsenabled make peer orderer configtxgen` as our doc says.

I have the feeling, until fabric core goes to a module, we might want to limit go modules to true chaincode?  But then maybe Matt has the answer to our troubles tomorrow ;-)

btw: in my case fpc was in the same GOPATH as fabric, as per (old) instructions ..

**Comment by g2flyer** [comment]
btw: what is also annoying is that each clean build seems to touch both go.mod and go.sum! I think the problems are the // indirect links in go.mod. We probably should do a go mod tidy before checking in go.mod when it contains indirect references?

**Comment by MHBauer** [comment]
> btw: i guess it's a go feature but it is surprising that now with modules a `go clean` triggered by `make clean` can trigger downloading of go modules ...?!

I've never even heard of `go clean` before. It would have to know what the result is, so it would have to have the modules it depends upon to know what to clean.

> Ok, just tried the PR on top of #307, did a `make clean all` and checked:

I can try this tomorrow.

> 
>     * does peer start with ecc-vscc enabled: answer no, failed launching peer with
> 
> ```
> panic: Error opening plugin at path ../../ecc/ecc-vscc.so: plugin.Open("../../ecc/ecc-vscc"): plugin was built with a different version of package github.com/hyperledger/fabric/common/errors
> ```

That seems especially strange. v2.0.1 should be v2.0.1 

>     * does peer start with ecc-vscc diabled: answer no, ercc installation fails with
> 
> 
> ```
> go: downloading github.com/hyperledger/fabric-chaincode-go v0.0.0-20200128192331-2d899240a7ed
> go: downloading github.com/hyperledger/fabric-protos-go v0.0.0-20200124220212-e9cfc186ba7b
> go: extracting github.com/hyperledger/fabric-protos-go v0.0.0-20200124220212-e9cfc186ba7b
> go: downloading github.com/golang/protobuf v1.3.2
> go: downloading google.golang.org/grpc v1.24.0
> go: extracting github.com/hyperledger/fabric-chaincode-go v0.0.0-20200128192331-2d899240a7ed
> go: extracting google.golang.org/grpc v1.24.0
> go: downloading google.golang.org/genproto v0.0.0-20190819201941-24fa4b261c55
> go: downloading golang.org/x/net v0.0.0-20200202094626-16171245cfb2
> go: downloading golang.org/x/sys v0.0.0-20191120155948-bd437916bb0e
> go: extracting github.com/golang/protobuf v1.3.2
> go: extracting golang.org/x/net v0.0.0-20200202094626-16171245cfb2
> go: downloading golang.org/x/text v0.3.2
> go: extracting golang.org/x/sys v0.0.0-20191120155948-bd437916bb0e
> go: extracting github.com/hyperledger/fabric v2.0.1+incompatible
> go: downloading go.uber.org/zap v1.10.0
> go: downloading github.com/sykesm/zap-logfmt v0.0.2
> go: downloading github.com/pkg/errors v0.8.1
> go: extracting google.golang.org/genproto v0.0.0-20190819201941-24fa4b261c55
> go: extracting github.com/pkg/errors v0.8.1
> go: extracting github.com/sykesm/zap-logfmt v0.0.2
> go: extracting go.uber.org/zap v1.10.0
> go: downloading go.uber.org/atomic v1.6.0
> go: downloading go.uber.org/multierr v1.5.0
> go: extracting golang.org/x/text v0.3.2
> go: extracting go.uber.org/multierr v1.5.0
> go: extracting go.uber.org/atomic v1.6.0
> go: updates to go.mod needed, disabled by -mod=readonly
> ```

~Yea, potentially I broke ercc somehow. I'm surprised that `-mod=readonly` is on in the ccenv.~ The readonly prompt comes from the buildcache being disabled because the filesystem is write only. Not sure if we can fix that, but I probably need to update ercc go.mod and then we need to do ercc and ecc test builds with -mod=readonly during CI.

> The `different version of package` error happened with both building fabric as `make native` as marcus said or based on `GO_TAGS=pluginsenabled make peer orderer configtxgen` as our doc says.

Not sure what the second option is.

> I have the feeling, until fabric core goes to a module, we might want to limit go modules to true chaincode? But then maybe Matt has the answer to our troubles tomorrow ;-)

I can potentially make a PR to fabric upstream. After the meeting tomorrow I will send an email to the mailing list about the existing JIRA issue;  https://jira.hyperledger.org/browse/FAB-10562.

> btw: in my case fpc was in the same GOPATH as fabric, as per (old) instructions ..

Shouldn't matter if it's on golang@v1.14

**Comment by MHBauer** [comment]
Is that also on top of #307? (I do not think it could impact this, but I
want to ensure when I test it that we all have the same environment).

**Comment by mbrandenburger** [comment]
> Is that also on top of #307? (I do not think it could impact this, but I want to ensure when I test it that we all have the same environment).
> […](#)
> On 3/27/20 3:58 AM, Marcus Brandenburger wrote: ***@***.**** commented on this pull request. Thanks Morgan! with this I can indeed build ecc-vscc but having the same problem as already mentioned by Michael when loading the plugin with Fabric. |panic: Error opening plugin at path /project/src/github.com/hyperledger-labs/fabric-private-chaincode/ecc/ecc-vscc.so: plugin.Open("/project/src/github.com/hyperledger-labs/fabric-private-chaincode/ecc/ecc-vscc"): plugin was built with a different version of package githu b.com/hyperledger/fabric/common/errors | — You are receiving this because you authored the thread. Reply to this email directly, view it on GitHub <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_hyperledger-2Dlabs_fabric-2Dprivate-2Dchaincode_pull_310-23pullrequestreview-2D382758091&d=DwMCaQ&c=jf_iaSHvJObTbx-siA1ZOg&r=gSVznMk4DixLVQ8USuNSH0RuT_4psswDPWAf-45M48c&m=NffWoRLQIn9ERt2GJUyvexeOmC6gGP7mrXbJc3MEyEk&s=YEHMJ7TSBJ_QIBZy57m0RWkSonXr5uyPoU-aQI4HSsI&e=>, or unsubscribe <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_ADFYGAOXZPDWR6DTKFXSHNDRJSBFHANCNFSM4LSZLGZQ&d=DwMCaQ&c=jf_iaSHvJObTbx-siA1ZOg&r=gSVznMk4DixLVQ8USuNSH0RuT_4psswDPWAf-45M48c&m=NffWoRLQIn9ERt2GJUyvexeOmC6gGP7mrXbJc3MEyEk&s=DTMeOu6JLWWrath1O69ewPQEA4i355vMnYhH0OVXiYY&e=>.

just this plain PR ... tested inside dev docker.

**Comment by g2flyer** [comment]
From what Matt said today, we might better postpone the go-mod-effort in FPC. Probably makes sense to keep ercc as mod (and maybe make ecc one?) but the rest probably stays better in old-style GOPATH mode?  Sorry, morgan, would mean of course you wasted a lot of effort for naught :-(

**Comment by MHBauer** [comment]
Huh, thought I had responded to this.

I agree that is makes sense to only use modules for the user-plugins.

**Comment by g2flyer** [comment]
given that fabric core is still not modularized, we table go mod-ification effort for now ...

**Pull request closed without merging**